Develop

awbarclay opened the pull request:
Hi everyone,

Before I get too far along in redoing the GCL R scripts, I was wondering if you could take a look at the functions I've worked on so far.  I'm not exactly sure how this pull request is going to work as I've never done one before.  Please check out functions with changes, there are 10. Most of the function are updates to the old functions but I did create a few new functions. You should be able to add review comments and comments in lines of code if you choose to commit changes. You can ignore the GTSeq_Metadata_Report.GCL function if you'd like, it's a function I created for Heather to pull a report from LOKI.

Thanks,
Andy

Commit:
Initial commit for old2new_gcl.GCL and old2new_LocCtrl.GCL.R.

Commit:
Made some corrections and bug fixes to old2new_LocCtrl.GCL and old2new_gcl.GCL after testing.

Commit:
initial commit for silly_n function.  Fixed a couple of inconsistencies old2new functions.

Commit:
added package name to p_load

Commit:
First commit for CreateLocusControl.GCL. This was tested using markersuite and locusnames and it works.

Commit:
as_tibble was giving a warning message about using the function on objects with more than two dimensions. Dropped the third dimension of the scores for each dose.

Commit:
Initial LOKI2R.GCL commit. Updated function documentation.

Commit:
Updated LOKI2R.GCL using tidy coding, need to test this version.

Commit:
Cleaned up LOKI2R.GCL and tested with GTSeq Chinook baseline. It pulled 157 collections for 557 loci in 4.5 min!

Commit:
Added workspace and changed sillyvec in example documentation for LOKI2R.GCL

Commit:
removed two unneeded variables from LOKI2R output

Commit:
Initial commit for tidy RemoveNAindiv.GCL. Need to test as function.

Commit:
Corrected documentation and output object for RemoveNAindiv.GCL

Commit:
Inital commit for tidy PoolCollections.GCL.  Only the documentation has been modified.

Commit:
Finished tidying up PoolCollections.GCL. This version needs to be tested.

Commit:
Made loci argument default to LocusControl$locusnames and added the ability to subset the pooled collection by a reduced set of loci.

Commit:
Removed the stop message to make sure that the IDs object is character in PoolCollections. Added documentation that says either character or numeric will work.

Commit:
Initial commit for RemoveIndMissLoci.GCL. This has been tested, but others should test as well.

Commit:
Added loci argument, documentation, and package names:: to functions in RemoveIndMissLOci.GCL

Commit:
added some package names to functions that were missed.

Commit:
initial commit for CheckDupWithinSilly.GCL. The documentation is updated but the code isn't ready to even test yet.

Commit:
CheckDupWithinSilly code updated for when quantile = NULL.Needs more testing.

Commit:
added mulitcoring to dupcheck function

Commit:
First commit for GTSeq_Metadata_Report function. Need to run this code as a function and verify with Heather H. that all required fields are in the output file.

Commit:
Cleaned up code in GTSeq_Metadata_Report

Commit:
Added quantile functionality to CheckDupWithinSilly.GCL. Also cleaned up code and added packages to non base functions. This version works line by line, but needs testing as a function.

Commit:
GTSeg_Metadata_Report.GCL-Added new view to database query and simplified it. The code works but needs to be tested as a function.

Commit:
minor fixes to dupcheck function

awbarclay commented:
In your review, you can focus on the changes that I've made to the code. If you have any ideas to improve the functions please add those to your comments.  Also, please suggest any changes that you would like made to the output of these functions or if there are any arguments you think I should add.  Please complete your review by next Wednesday (May20).  Hopefully that will give you enough time.

emlee-ak commented on old2new_LocCtrl.GCL.R:
```diff
@@ -0,0 +1,53 @@
+    
+    tibble::tibble(allele = seq(length(a)), call = a)
+    
+  }) %>% purr::set_names(LocusControl$locusnames)
```
Change to "purrr".

Error in loadNamespace(name) : there is no package called ‘purr’

awbarclay replied:
Good catch, I'll fix it. I keep forgetting the third "r".

emlee-ak commented on old2new_gcl.GCL.R:
```diff
@@ -0,0 +1,61 @@
+  # Inputs~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  #   sillyvec - A character vector of silly codes with associated *.gcl objects that need to be converted.
+  #   overwrite - logical; whether you want the old objects overwritten without saving 
+  #              (TRUE) or assign the new objects to *.gcl_old before overwriting (FALSE) 
```
Always overwrites old .gcl object regardless of "overwrite" logical value. No *.gcl_old object created with "FALSE" or empty.

Commit:
Fixed issue with single fish collections causing the function to bomb.

Commit:
Correctd line 38 and added line to make sure column names are correct.

krshedd committed:
Minor change to old2new_LocCtrl
Included package for tibble and removed purrr by switching to sapply

emlee-ak commented on silly_n.GCL.R:
```diff
@@ -0,0 +1,34 @@
+  # load("V:/Analysis/2_Central/Chinook/Cook Inlet/2019/2019_UCI_Chinook_baseline_hap_data/2019_UCI_Chinook_baseline_hap_data.RData")
+  #
+  # SampleSize.GCL(sillyvec = sillyvec157)
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Change function name in example to silly_n.GCL()

Commit:
old2new_gcl.GCL - added a line to make sure SILLY_CODE is included in the attributes. This was added because the attributes in older analyses may not include this column.

Commit:
Merge branch 'develop' of https://github.com/krshedd/GCL-R-Scripts into develop

krshedd committed:
Minor changes to old2new_gcl
Included if/stop to check for missing sillys in workspace, package for mutate, add line break for final message

krshedd committed:
Minor changes to silly_n
Updated example with correct function name, package for bind_rows

krshedd committed:
Minor change old2new_LocCtrl
Indented brackets appropriately for if statements

krshedd committed:
old2new_gcl no extra lines at end

krshedd committed:
Minor changes to CreateLocusControl
Standardize indentation with Ctrl + i, change lapply with setnames to sapply with simplify = FALSE

krshedd committed:
Bugfix CreateLocusControl, simplify = FALSE

krshedd committed:
Minor changes to LOKI2R
Standardize indentation with Ctrl + i, specify packages for non-base functions

krshedd commented on LOKI2R.GCL.R:
```diff
@@ -52,132 +84,126 @@ LOKI2R.GCL=function(sillyvec,username,password){
-
-    }
       
-    scores <- array(data=NA,dim=c(nind,length(loci),max(ploidy)),dimnames=list(ids,loci,paste0("Dose",seq(max(ploidy)))))
```
Hey Andy, I definitely dig the new tibble format for .gcl objects, but one thing I did notice since we are no longer filling in a pre-defined array, sillys that don't have data for certain loci no longer have for NA's for those loci in the final tibble, they just don't have those loci (columns) at all in the final tibble. Currently there is no way that a user would know that the collection didn't have genotypes for all loci until you try `RemoveNAindiv.GCL`, and you'd only know then because the function bombs because it can't find those loci to see if there are NAs. Make sense? 

How should we deal with this? Based on our current workflows, I think that each silly should probably have columns for all loci in LocusControl and fill in NAs for any loci without genotypes, but I'm open to other ideas.

krshedd replied:
We could also toss out a message that says that loci are missing for some collections and automatically dump out a tibble showing which sillys are missing which loci.

awbarclay replied:
I think this is a good idea.  That way you know right away if some of the sillys your are trying to pull have missing data.  Sillys without any data will not get pulled from LOKI at all because they don't exist in the lookup table.  I can add those to the missing data tibble as well.

krshedd commented on RemoveNAindv.GCL.R:
```diff
@@ -1,47 +1,54 @@
+  #    2) output is tibble of the fish IDs removed by silly
+  #
+  # Example~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  #   load("V:/Analysis/2_Central/Chinook/Cook Inlet/2019/2019_UCI_Chinook_baseline_hap_data/2019_UCI_Chinook_baseline_hap_data.RData")
+  # 
+  #   removedInd <- RemoveNAindv.GCL(sillyvec = sillyvec157)
```
This example doesn't really work, we should find an example that has results (i.e. individuals removed), also all of the .gcl objects and LocusControl are in the old list format, so we should add a complete, reproducible example.

awbarclay replied:
I forgot to change the workspace image to one that has the new style object, I'll change the directory.  As for being reproducible, should our example data be stored on GitHub somehow or maybe as a data object in our GCL package so the example will work even if you aren't on the dfg network?

krshedd replied:
I've rolled the remove NA individual process in to LOKI2R, I think we can probably get rid of this function now, what do you think?

krshedd replied:
Actually, on second thought, maybe not. If someone converts an old *.gcl list object to the new format, they may still need to remove NA individuals. This function just isn't necessary for dealing with new pulls via LOKI2R

krshedd committed:
Formating change to RemoveNAindiv
Still needs to be fully tested

awbarclay commented:
Hi Kyle,

Good catch.  I’ve only been testing these function on sillys with data for all loci so I didn’t notice.  I like your idea of making sure that each .gcl contains a column for all loci in LocusControl even if they don’t have scores. I’ll make some changes to LOKI2R.GCL and old2new_gcl push the changes.

Andy

From: Kyle Shedd <notifications@github.com>
Sent: Wednesday, May 20, 2020 4:40 PM
To: krshedd/GCL-R-Scripts <GCL-R-Scripts@noreply.github.com>
Cc: Barclay, Andy W (DFG) <andy.barclay@alaska.gov>; Assign <assign@noreply.github.com>
Subject: Re: [krshedd/GCL-R-Scripts] Develop (#34)


@krshedd commented on this pull request.

________________________________

In LOKI2R.GCL.R<https://urldefense.com/v3/__https:/github.com/krshedd/GCL-R-Scripts/pull/34*discussion_r428384455__;Iw!!J2_8gdp6gZQ!6R_I0F6GPLbORk9EMjzgNCz9weXGpAxH8sXwx75hu9LodgkEjtUniD3r1fXC4bEEZNa3Hw$>:

>

-    scores <- array(data=NA,dim=c(nind,length(loci),max(ploidy)),dimnames=list(ids,loci,paste0("Dose",seq(max(ploidy)))))

Hey Andy, I definitely dig the new tibble format for .gcl objects, but one thing I did notice since we are no longer filling in a pre-defined array, sillys that don't have data for certain loci no longer have for NA's for those loci in the final tibble, they just don't have those loci (columns) at all in the final tibble. Currently there is no way that a user would know that the collection didn't have genotypes for all loci until you try RemoveNAindiv.GCL, and you'd only know then because the function bombs because it can't find those loci to see if there are NAs. Make sense?

How should we deal with this? Based on our current workflows, I think that each silly should probably have columns for all loci in LocusControl and fill in NAs for any loci without genotypes, but I'm open to other ideas.

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/krshedd/GCL-R-Scripts/pull/34*pullrequestreview-415806478__;Iw!!J2_8gdp6gZQ!6R_I0F6GPLbORk9EMjzgNCz9weXGpAxH8sXwx75hu9LodgkEjtUniD3r1fXC4bGt7Tayqg$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AEJYSJC4JIMMEUKHNIPBPSTRSRZ7TANCNFSM4NAFUHFQ__;!!J2_8gdp6gZQ!6R_I0F6GPLbORk9EMjzgNCz9weXGpAxH8sXwx75hu9LodgkEjtUniD3r1fXC4bFBpj1Tpw$>.

Commit:
Added empty scores columns to .gcl objects and updated example.

Commit:
Added messages and tibble for sillys with missing data.

krshedd commented on LOKI2R.GCL.R:
```diff
@@ -52,132 +87,157 @@ LOKI2R.GCL=function(sillyvec,username,password){
-
-  }#silly 
-
+    readr::write_csv(missing_loci, path = "LOKI2R_missing_loci.csv")
```
I like the idea of returning missing loci to the environment, but I guess I was leaning more towards assigning it to an object in your workspace, rather than writing out a file to your working directory.

krshedd committed:
Change LOKI2R to assign rather than write missing_loci
Also changed messages to warnings (just in case people run code chunks with message=FALSE), simplified some code, and explicitly mentioned package names

krshedd commented on RemoveNAindv.GCL.R:
```diff
@@ -1,47 +1,60 @@
-      scores <- scores[ind.to.keep, , ,drop=FALSE]
-      n <- nrow(scores)
-      attributes <- attributes[ind.to.keep, ,drop=FALSE]
+      sum(is.na(ind)) > 0 
       
-      assign(paste(silly, ".gcl", sep = ""), list(counts = counts, scores = scores, n = n, attributes = attributes), pos = 1)
-    }} )
+    })]
```
So this actually ends up dropping any individual that is missing data for any reason (no calls and never genotyped), because we convert all missing genotypes to NA in LOKI2R. In the old format of .gcl objects, the scores array had NAs for never genotyped and 0s for no calls. This function would then look for individuals with NAs (i.e. never genotyped for a given locus). Maybe we should just retire this function and fold it in with LOKI2R since we already calculate missing_loci? We can use that to cull individuals that were not genotyped for all loci.

krshedd replied:
I figured out what the problem was here, we were looking at allele_1 and allele_2. All haploid/mitochondrial markers are default NA for allele_2, so if your markersuite contains any haploid markers, all individuals are removed. Just look at allele_1.

krshedd committed:
Trivial change to RemoveNAindv, still not working

krshedd committed:
Bugfix RemoveNAindv, only look for NA's in allele_1
Mitochondrial or haploid markers are NA for all allele_2, so if we look in allele_2 as well, all fish are removed if haploid markers exist

krshedd committed:
LOKI2R - remove NA indv, replace "0" with NA
Also updated the example and specified packages for non-base functions

krshedd committed:
Trivial changes to PoolCollections

krshedd committed:
Bugfix RemoveIndMissLoci to find NA, not 0

krshedd committed:
Update LOKI2R, name missing_indvs_loci$missing_loci with SillySource for easy indexing with $

krshedd commented on CheckDupWithinSilly.GCL.r:
```diff
@@ -1,127 +1,281 @@
+        
+        dimnames(scores1)[[1]] <- IDs
+        
+        sort.scores.df <- reshape::sort_df(scores1)
+        
+        sortIDs <- dimnames(sort.scores.df)[[1]]
+        
+     #Calculate the porpotion of duplicated loci for each silly
+      duplication <- t(sapply(1:(n-1), function(id){ 
+            
+        vec <- sapply(1:nloci, function(locus){ 
+          
+          sort.scores.df[id, locus]==sort.scores.df[id+1, locus]
+          
+        });
+        
+        ifelse(sum(!is.na(vec)), sum(vec[!is.na(vec)])/sum(!is.na(vec)), 0 )
+        
+      }))
```
This is a bug, after sorting the entire data.frame, it only looks for duplicate fish on either side. If two duplicate fish have NA's at different loci, they may not sort next to each other, and thus may be missed by this lazy look.

krshedd replied:
This bug has been in here from the beginning, from the looks of it.

krshedd replied:
You can test this yourself if you run the function with `quantile = NULL` and with `quantile = 0.99` and you can end up with different numbers of duplicates.

krshedd committed:
Update CheckDupWithinSilly + Found a bug
Clarified some of the duplicate check code to optimize for speed, added package names, formating. Also found a bug where we miss duplicates if quantile = NULL, because NA genotypes can cause the lazy sorting to miss near duplicates. This bug is not yet resolved.

Commit:
CheckDupWithinSilly - change code for when quantile=NULL to use rubias::close_matching_samples() This needs further testing after the update.

Commit:
CheckDupWithinSilly - added new documentation and fixed a couple of issues causing the function to bomb when qantile=NULL

Commit:
CheckDupWithinSilly - Replaced old example in documentation

Commit:
CheckDupWithinSilly - minor changes to documentation

Commit:
CheckDupWithinSilly - added stop message for individuals missing all loci.

Commit:
CheckDupsWithinSilly - added workaround for hapliod markers when using quantile=NULL  and some other minor changes for consistency between the two methods.

Commit:
CheckDupWithinSilly - made a minor fix to my.gcl

krshedd committed:
Minor updates to CheckDupWithinSilly, default quantile = NULL

Also clarified some code, added package names, formatting
